etcdctlv3/command/snapshot_command.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2016 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package command
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/coreos/etcd/Godeps/_workspace/src/github.com/spf13/cobra"
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/coreos/etcd/clientv3"
+)
+
+// NewSnapshotCommand returns the cobra command for "snapshot".
+func NewSnapshotCommand() *cobra.Command {
+	return &cobra.Command{
+		Use:   "snapshot [filename]",
+		Short: "Snapshot streams a point-in-time snapshot of the store",
+		Run:   snapshotCommandFunc,
+	}
+}
+
+// snapshotCommandFunc watches for the length of the entire store and records
+// to a file.
+func snapshotCommandFunc(cmd *cobra.Command, args []string) {
+	switch {
+	case len(args) == 0:
+		snapshotToStdout(clientv3.NewWatcher(mustClient(cmd)))
+	case len(args) == 1:
+		snapshotToFile(clientv3.NewWatcher(mustClient(cmd)), args[0])
+	default:
+		err := fmt.Errorf("snapshot takes at most one argument")
+		ExitWithError(ExitBadArgs, err)
+	}
+}
+
+// snapshotToStdout streams a snapshot over stdout
+func snapshotToStdout(wapi clientv3.Watcher) {
+	wch := wapi.WatchPrefix(context.TODO(), "", 1)
+	if wr := <-wch; len(wr.Events) > 0 {
+		// no compactions in history; start writing snapshot
+		if !printEvents(os.Stdout, wr) {
+			return
+		}
+	} else {
+		// retry with latest compaction
+		wch = wapi.WatchPrefix(context.TODO(), "", wr.CompactRevision)
```

Comment:
OK. I don't think `Range()` can do that now; querying with [a, b) means there can be a c > b that won't be in the interval. API bug?